OrbitQt/ConnectToStadiaWidget.cpp

```diff
@@ -0,0 +1,344 @@
+            }
+
+            emit ReceivedSshInfo();
+          });
+    }
+  });
+}
+
+void ConnectToStadiaWidget::CheckCredentialsAvailable() {
+  CHECK(connection_artifacts_->selected_instance_);
+
+  const std::string& instance_id = connection_artifacts_->selected_instance_->id.toStdString();
+
+  if (!instance_credentials_.contains(instance_id)) return;
+
+  if (instance_credentials_.at(instance_id).has_error()) {
+    emit ErrorOccurred(
+        QString::fromStdString(instance_credentials_.at(instance_id).error().message()));
+    return;
+  }
+
+  emit ReadyToDeploy();
+}
+
+void ConnectToStadiaWidget::DeployOrbitService() {
+  CHECK(connection_artifacts_->service_deploy_manager_ == nullptr);
+  CHECK(connection_artifacts_->selected_instance_);
+  const std::string instance_id = connection_artifacts_->selected_instance_->id.toStdString();
+  CHECK(instance_credentials_.contains(instance_id));
+  CHECK(instance_credentials_.at(instance_id).has_value());
+
+  const OrbitSsh::Credentials& credentials{instance_credentials_.at(instance_id).value()};
+
+  connection_artifacts_->CreateServiceDeployManager(credentials);
+
+  ScopedConnection label_connection{
+      QObject::connect(connection_artifacts_->service_deploy_manager_.get(),
+                       &orbit_qt::ServiceDeployManager::statusMessage, ui_->instancesTableOverlay,
+                       &OverlayWidget::SetStatusMessage)};
+  ScopedConnection cancel_connection{
+      QObject::connect(ui_->instancesTableOverlay, &OverlayWidget::Cancelled,
+                       connection_artifacts_->service_deploy_manager_.get(),
+                       &orbit_qt::ServiceDeployManager::Cancel)};
+
+  const auto deployment_result = connection_artifacts_->service_deploy_manager_->Exec();
+  if (!deployment_result) {
+    Disconnect();
+    if (deployment_result.error() ==
+        make_error_code(orbit_qt::Error::kUserCanceledServiceDeployment)) {
+      return;
+    }
+    emit ErrorOccurred(QString("Orbit was unable to successfully connect to the Instance. The "
+                               "error message was: %1")
+                           .arg(QString::fromStdString(deployment_result.error().message())));
+    return;
+  }
+
+  LOG("Deployment successful, grpc_port: %d", deployment_result.value().grpc_port);
+  CHECK(connection_artifacts_->grpc_channel_ == nullptr);
+  std::string grpc_server_address =
+      absl::StrFormat("127.0.0.1:%d", deployment_result.value().grpc_port);
+  LOG("Starting gRPC channel to: %s", grpc_server_address);
+  connection_artifacts_->grpc_channel_ = grpc::CreateCustomChannel(
+      grpc_server_address, grpc::InsecureChannelCredentials(), grpc::ChannelArguments());
+  CHECK(connection_artifacts_->grpc_channel_ != nullptr);
+
+  emit Connected();
+}
+
+void ConnectToStadiaWidget::Disconnect() {
+  connection_artifacts_->grpc_channel_ = nullptr;
+
+  // TODO(antonrohr) currently does not work
```

Comment:
Can you link a bug to the 2do please? I didn't get around to look into it yet.

Reply:
done